In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
#import  as
from sklearn.preprocessing import LabelEncoder, RobustScaler, StandardScaler, MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report


In [2]:
data = pd.read_csv("E:\ML\AISS\hotel_booking_3_3010\hotel_bookings.csv")

In [3]:
data.head(2)

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type  adr  \
0    NaN     NaN                    0     Transient  0.0   
1    NaN     NaN                    0     Transient  0.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  

[2 rows x 32 columns]

In [4]:
data.shape
data.hotel.unique()

array(['Resort Hotel', 'City Hotel'], dtype=object)

In [6]:
#data.info()
data.dtypes

hotel                              object
is_canceled                         int64
lead_time                           int64
arrival_date_year                   int64
arrival_date_month                 object
arrival_date_week_number            int64
arrival_date_day_of_month           int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                               object
country                            object
market_segment                     object
distribution_channel               object
is_repeated_guest                   int64
previous_cancellations              int64
previous_bookings_not_canceled      int64
reserved_room_type                 object
assigned_room_type                 object
booking_changes                     int64
deposit_type                       object
agent                             

In [7]:
#data.hist()
fig_dims = (50, 50)
fig, ax = plt.subplots(figsize=fig_dims)
hist = data.hist(bins=20, ax=ax) 
plt.savefig("hist.png")
#plt.show()
plt.close()

c:\users\sheetal\appdata\local\programs\python\python37\lib\site-packages\IPython\core\interactiveshell.py:3296: UserWarning: To output multiple subplots, the figure containing the passed axes is being cleared
  exec(code_obj, self.user_global_ns, self.user_ns)


In [8]:
data.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

In [9]:
del data["company"]

In [10]:
data.shape

(119390, 31)

In [11]:
data.corr()

is_canceled  lead_time  arrival_date_year  \
is_canceled                        1.000000   0.293123           0.016660   
lead_time                          0.293123   1.000000           0.040142   
arrival_date_year                  0.016660   0.040142           1.000000   
arrival_date_week_number           0.008148   0.126871          -0.540561   
arrival_date_day_of_month         -0.006130   0.002268          -0.000221   
stays_in_weekend_nights           -0.001791   0.085671           0.021497   
stays_in_week_nights               0.024765   0.165799           0.030883   
adults                             0.060017   0.119519           0.029635   
children                           0.005048  -0.037622           0.054624   
babies                            -0.032491  -0.020915          -0.013192   
is_repeated_guest                 -0.084793  -0.124410           0.010341   
previous_cancellations             0.110133   0.086042          -0.119822   
previous_bookings_not_canceled    -0.057358  -0.073548           0.029218   
booking_changes                   -0.144381   0.000149           0.030872   
agent                             -0.083114  -0.069741           0.063457   
days_in_waiting_list               0.054186   0.170084          -0.056497   
adr                                0.047557  -0.063077           0.197580   
required_car_parking_spaces       -0.195498  -0.116451          -0.013684   
total_of_special_requests         -0.234658  -0.095712           0.108531   

                                arrival_date_week_number  \
is_canceled                                     0.008148   
lead_time                                       0.126871   
arrival_date_year                              -0.540561   
arrival_date_week_number                        1.000000   
arrival_date_day_of_month                       0.066809   
stays_in_weekend_nights                         0.018208   
stays_in_week_nights                            0.015558   
adults                                          0.025909   
children                                        0.005518   
babies                                          0.010395   
is_repeated_guest                              -0.030131   
previous_cancellations                          0.035501   
previous_bookings_not_canceled                 -0.020904   
booking_changes                                 0.005508   
agent                                          -0.031201   
days_in_waiting_list                            0.022933   
adr                                             0.075791   
required_car_parking_spaces                     0.001920   
total_of_special_requests                       0.026149   

                                arrival_date_day_of_month  \
is_canceled                                     -0.006130   
lead_time                                        0.002268   
arrival_date_year                               -0.000221   
arrival_date_week_number                         0.066809   
arrival_date_day_of_month                        1.000000   
stays_in_weekend_nights                         -0.016354   
stays_in_week_nights                            -0.028174   
adults                                          -0.001566   
children                                         0.014544   
babies                                          -0.000230   
is_repeated_guest                               -0.006145   
previous_cancellations                          -0.027011   
previous_bookings_not_canceled                  -0.000300   
booking_changes                                  0.010613   
agent                                            0.001487   
days_in_waiting_list                             0.022728   
adr                                              0.030245   
required_car_parking_spaces                      0.008683   
total_of_special_requests                        0.003062   

                                stays_in_weekend_nights  stays_in_wee

In [12]:
print(data.country.mode())
data.country = data.country.fillna("PRT")
print(data.agent.mean())
data.agent = data.agent.fillna(86.69338185346918)
print(data.children.mode())
data.children = data.children.fillna(0.0)
# print(data.meal.mode())
# data.meal = data.meal.fillna("BB")
# print(data.market_segment.mode())
# data.market_segment = data.market_segment.fillna("Online TA")
# print(data.distribution_channel.mode())
# data.distribution_channel = data.distribution_channel.fillna("TA/TO")

0    PRT
dtype: object
86.69338185346919
0    0.0
dtype: float64


In [13]:
data.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces 

In [14]:
data.previous_cancellations.unique()
data.agent.unique()
data.assigned_room_type.unique()
data.booking_changes.unique()
data.reserved_room_type.unique()
data.previous_bookings_not_canceled.unique()
data.distribution_channel.unique()
data.market_segment.unique()
# data.country.unique()
# data.meal.unique()
# data.is_canceled.unique()
# data.arrival_date_year.unique()
# data.arrival_date_month.unique()
# data.arrival_date_week_number.unique()
# data.stays_in_week_nights.unique()
# data.stays_in_weekend_nights.unique()
# data.adults.unique()
# data.children.unique()
# data.babies.unique()
# data.is_repeated_guest.unique()
# data.previous_cancellations.unique()
# data.deposit_type.unique()
# data.days_in_waiting_list.unique()
# data.customer_type.unique()
# data.adr.unique()
# data.required_car_parking_spaces.unique()
# data.total_of_special_requests.unique()
# data..unique()

array(['Direct', 'Corporate', 'Online TA', 'Offline TA/TO',
       'Complementary', 'Groups', 'Undefined', 'Aviation'], dtype=object)

In [15]:
data.describe()

is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2016.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2015.000000   
25%         0.000000      18.000000        2016.000000   
50%         0.000000      69.000000        2016.000000   
75%         1.000000     160.000000        2017.000000   
max         1.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_week_nights         adults  \
count            119390.000000         119390.000000  119390.000000   
mean                  0.927599              2.500302       1.856403   
std                   0.998613              1.908286       0.579261   
min                   0.000000              0.000000       0.000000   
25%                   0.000000              1.000000       2.000000   
50%                   1.000000              2.000000       2.000000   
75%                   2.000000              3.000000       2.000000   
max                  19.000000             50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119390.000000  119390.000000      119390.000000   
mean        0.103886       0.007949           0.031912   
std         0.398555       0.097436           0.175767   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119390.000000                   119390.000000   
mean                 0.087118                        0.137097   
std                  0.844336                        1.497437   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes          agent  days_in_waiting_list            adr  \
count    119390.000000  119390.000000         119390.000000  119390.000000   
mean          0.221124      86.693382              2.321149     101.831122   
std           0.652306     102.915247             17.594721      50.535790   
min           0.000000       1.000000              0.000000      -6.380000   
25%           0.000000       9.000000              0.000000      69.290000   
50%           0.000000      28.000000              0.000000      94.575000   
75%           0.000000     152.000000              0.000000     126.000000   
max          21.000000     535.000000            391.000000    5400.000000   

       required_car_parking_spaces  total_of_special_requests  
count                119390.000000              119390.000000  
mean                      0.062518                   0.571363  
std                       0.245291                   0.792798  
min                       0.000000                   0.000000  
25%                       0.000000                   0.000000  
50%                       0.000000                   0.000000  
75%                       0

In [16]:
#data.boxplot()
fig_dims = (150, 150)
fig, ax = plt.subplots(figsize=fig_dims)
box = data.boxplot() 
plt.savefig("box.png")
#plt.show()
plt.close()

In [17]:
scaler = RobustScaler()
#scaler = StandardScaler()
#scaler = MinMaxScaler()
data[["lead_time","stays_in_weekend_nights","stays_in_week_nights","adults",
      "children","babies","previous_cancellations","previous_bookings_not_canceled",
      "booking_changes","days_in_waiting_list","adr","required_car_parking_spaces",
      "agent","total_of_special_requests"]] = \
scaler.fit_transform(data[["lead_time","stays_in_weekend_nights","stays_in_week_nights",
                           "adults","children","babies","previous_cancellations",
                           "previous_bookings_not_canceled","booking_changes","days_in_waiting_list",
                           "adr","required_car_parking_spaces","agent","total_of_special_requests"]])

#lead_time	arrival_date_year	arrival_date_week_number	arrival_date_day_of_month	
#stays_in_weekend_nights	stays_in_week_nights	
#adults	children	babies	is_repeated_guest	previous_cancellations	
#previous_bookings_not_canceled	booking_changes	agent	days_in_waiting_list	
#adr	required_car_parking_spaces	total_of_special_requests

In [18]:
label_encoder = LabelEncoder()

In [19]:
data.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0   1.922535               2015               July   
1  Resort Hotel            0   4.704225               2015               July   
2  Resort Hotel            0  -0.436620               2015               July   
3  Resort Hotel            0  -0.394366               2015               July   
4  Resort Hotel            0  -0.387324               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
0                     -0.5                  -1.0     0.0  ...   
1                     -0.5                  -1.0     0.0  ...   
2                     -0.5                  -0.5    -1.0  ...   
3                     -0.5                  -0.5    -1.0  ...   
4                     -0.5                   0.0     0.0  ...   

   booking_changes  deposit_type     agent days_in_waiting_list customer_type  \
0              3.0    No Deposit  0.410443                  0.0     Transient   
1              4.0    No Deposit  0.410443                  0.0     Transient   
2              0.0    No Deposit  0.410443                  0.0     Transient   
3              0.0    No Deposit  1.930070                  0.0     Transient   
4              0.0    No Deposit  1.482517                  0.0     Transient   

        adr  required_car_parking_spaces  total_of_special_requests  \
0 -1.667695                          0.0                        0.0   
1 -1.667695                          0.0                        0.0   
2 -0.345177                          0.0                        0.0   
3 -0.345177                          0.0                        0.0   
4  0.060395                          0.0                        1.0   

   reservation_status reservation_status_date  
0           Check-Out              2015-07-01  
1           Check-Out              2015-07-01  
2           Check-Out              2015-07-02  
3           Check-Out              2015-07-02  
4           Check-Out              2015-07-03  

[5 rows x 31 columns]

In [20]:
data['hotel']= label_encoder.fit_transform(data['hotel']) 
data['arrival_date_year']= label_encoder.fit_transform(data['arrival_date_year']) 
data['arrival_date_month']= label_encoder.fit_transform(data['arrival_date_month']) 
data['meal']= label_encoder.fit_transform(data['meal']) 
data['country']= label_encoder.fit_transform(data['country']) 
data['market_segment']= label_encoder.fit_transform(data['market_segment']) 
data['distribution_channel']= label_encoder.fit_transform(data['distribution_channel']) 
data['reserved_room_type']= label_encoder.fit_transform(data['reserved_room_type']) 
data['assigned_room_type']= label_encoder.fit_transform(data['assigned_room_type']) 
data['deposit_type']= label_encoder.fit_transform(data['deposit_type']) 
data['customer_type']= label_encoder.fit_transform(data['customer_type']) 
#data['']= label_encoder.fit_transform(data['']) 
data['is_canceled']= label_encoder.fit_transform(data['is_canceled']) 
#data['arrival_date_week_number']= label_encoder.fit_transform(data['arrival_date_week_number']) 
data['children']= label_encoder.fit_transform(data['children']) 
data['reservation_status_date']= label_encoder.fit_transform(data['reservation_status_date']) 
data['reservation_status']= label_encoder.fit_transform(data['reservation_status']) 

In [21]:
data.describe()

hotel    is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000  119390.000000      119390.000000   
mean        0.335539       0.370416       0.246559           1.156554   
std         0.472181       0.482918       0.752557           0.707476   
min         0.000000       0.000000      -0.485915           0.000000   
25%         0.000000       0.000000      -0.359155           1.000000   
50%         0.000000       0.000000       0.000000           1.000000   
75%         1.000000       1.000000       0.640845           2.000000   
max         1.000000       1.000000       4.704225           2.000000   

       arrival_date_month  arrival_date_week_number  \
count       119390.000000             119390.000000   
mean             5.490544                 27.165173   
std              3.535075                 13.605138   
min              0.000000                  1.000000   
25%              2.000000                 16.000000   
50%              6.000000                 28.000000   
75%              8.000000                 38.000000   
max             11.000000                 53.000000   

       arrival_date_day_of_month  stays_in_weekend_nights  \
count              119390.000000            119390.000000   
mean                   15.798241                -0.036201   
std                     8.780829                 0.499307   
min                     1.000000                -0.500000   
25%                     8.000000                -0.500000   
50%                    16.000000                 0.000000   
75%                    23.000000                 0.500000   
max                    31.000000                 9.000000   

       stays_in_week_nights         adults  ...  booking_changes  \
count         119390.000000  119390.000000  ...    119390.000000   
mean               0.250151      -0.143597  ...         0.221124   
std                0.954143       0.579261  ...         0.652306   
min               -1.000000      -2.000000  ...         0.000000   
25%               -0.500000       0.000000  ...         0.000000   
50%                0.000000       0.000000  ...         0.000000   
75%                0.500000       0.000000  ...         0.000000   
max               24.000000      53.000000  ...        21.000000   

        deposit_type          agent  days_in_waiting_list  customer_type  \
count  119390.000000  119390.000000         119390.000000  119390.000000   
mean        0.124893       0.410443              2.321149       2.137323   
std         0.334678       0.719687             17.594721       0.577040   
min         0.000000      -0.188811              0.000000       0.000000   
25%         0.000000      -0.132867              0.000000       2.000000   
50%         0.000000       0.000000              0.000000       2.000000   
75%         0.000000       0.867133              0.000000       2.000000   
max         2.000000       3.545455            391.000000       3.000000   

                 adr  required_car_parking_spaces  total_of_special_requests  \
count  119390.000000                119390.000000              119390.000000   
mean        0.127951                     0.062518                   0.571363   
std         0.891127                     0.245291                   0.792798   
min        -1.780197                     0.000000                   0.000000   
25%        -0.445865                     0.000000                   0.000000   
50%         0.000000                     0.000000                   0.000000   
75%         0.554135                     0.000000                   1.000000   
max        93.553606                     8.000000                   5.000000   

       reservation_status  reservation_status_date  
count       119390.000000            119390.000000  
mean             0.649803               516.750147  
std              0.497776               227.476419  
min              0.000000                 0.000000  
25%              0.000000  

In [22]:
X = data.drop("is_canceled", axis=1)
X.shape
X.head()

hotel  lead_time  arrival_date_year  arrival_date_month  \
0      1   1.922535                  0                   5   
1      1   4.704225                  0                   5   
2      1  -0.436620                  0                   5   
3      1  -0.394366                  0                   5   
4      1  -0.387324                  0                   5   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  ...  \
0                     -0.5                  -1.0     0.0         0  ...   
1                     -0.5                  -1.0     0.0         0  ...   
2                     -0.5                  -0.5    -1.0         0  ...   
3                     -0.5                  -0.5    -1.0         0  ...   
4                     -0.5                   0.0     0.0         0  ...   

   booking_changes  deposit_type     agent  days_in_waiting_list  \
0              3.0             0  0.410443                   0.0   
1              4.0             0  0.410443                   0.0   
2              0.0             0  0.410443                   0.0   
3              0.0             0  1.930070                   0.0   
4              0.0             0  1.482517                   0.0   

   customer_type       adr  required_car_parking_spaces  \
0              2 -1.667695                          0.0   
1              2 -1.667695                          0.0   
2              2 -0.345177                          0.0   
3              2 -0.345177                          0.0   
4              2  0.060395                          0.0   

   total_of_special_requests  reservation_status  reservation_status_date  
0                        0.0                   1                      121  
1                        0.0                   1                      121  
2                        0.0                   1                      122  
3                        0.0                   1                      122  
4                        1.0                   1                      123  

[5 rows x 30 columns]

In [23]:
#data['is_canceled'] = data['is_canceled'].astype(float) 

In [1]:
y = data.is_canceled
y.shape
y.head()
X.head()

NameError: name 'data' is not defined

In [25]:
X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.2, random_state=0)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape, X.shape, y.shape)

(95512, 30) (23878, 30) (95512,) (23878,) (119390, 30) (119390,)


### RFA 99.99

In [26]:

classifier = RandomForestClassifier(n_estimators = 50)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

# calculate accuracy
print("\n\nConfusion Matrix: \n\n", confusion_matrix(y_test, y_pred))
print ("\n\nAccuracy : \n\n", accuracy_score(y_test,y_pred)*100)
print("\n\nReport : \n\n", classification_report(y_test, y_pred))



Confusion Matrix: 

 [[14934     0]
 [    0  8944]]


Accuracy : 

 100.0


Report : 

               precision    recall  f1-score   support

           0       1.00      1.00      1.00     14934
           1       1.00      1.00      1.00      8944

   micro avg       1.00      1.00      1.00     23878
   macro avg       1.00      1.00      1.00     23878
weighted avg       1.00      1.00      1.00     23878



# Confusion matrix comparison

Confusion Matrix: RFA
 [[14934    0]
  [    0  8944]]
  
 Confusion Matrix: Logistic Regression
 [[14931     3]
 [  259  8685]]
 
 Confusion Matrix: DT gini index
 [[14934     0]
 [    0  8944]]


Confusion Matrix: DT Entropy
 [[14934     0]
 [    0  8944]]

Confusion Matrix: GaussianNB
 [[14764   170]
 [    0  8944]]
 
 Confusion Matrix: adaboost
 [[14931     3]
 [  259  8685]]
 
 Confusion Matrix: xgb
 [[14934     0]
 [    0  8944]]

### Linear Regression
* only for binary values

### Logistic Regression

In [29]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

# calculate accuracy
print("\n\nConfusion Matrix: \n\n", confusion_matrix(y_test, y_pred))
print ("\n\nAccuracy : \n\n", accuracy_score(y_test,y_pred)*100)
print("\n\nReport : \n\n", classification_report(y_test, y_pred))

c:\users\sheetal\appdata\local\programs\python\python37\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)




Confusion Matrix: 

 [[14931     3]
 [  259  8685]]


Accuracy : 

 98.90275567467963


Report : 

               precision    recall  f1-score   support

           0       0.98      1.00      0.99     14934
           1       1.00      0.97      0.99      8944

   micro avg       0.99      0.99      0.99     23878
   macro avg       0.99      0.99      0.99     23878
weighted avg       0.99      0.99      0.99     23878



### Decision Tree gini index

In [48]:
from sklearn.tree import DecisionTreeClassifier
dt_gini = DecisionTreeClassifier(criterion = "gini", random_state = 100,max_depth=3, min_samples_leaf=5) 
dt_gini.fit(X_train, y_train)
y_pred = dt_gini.predict(X_test)

# calculate accuracy
print("\n\nConfusion Matrix: \n\n", confusion_matrix(y_test, y_pred))
print ("\n\nAccuracy : \n\n", accuracy_score(y_test,y_pred)*100)
print("\n\nReport : \n\n", classification_report(y_test, y_pred))



Confusion Matrix: 

 [[14934     0]
 [    0  8944]]


Accuracy : 

 100.0


Report : 

               precision    recall  f1-score   support

           0       1.00      1.00      1.00     14934
           1       1.00      1.00      1.00      8944

   micro avg       1.00      1.00      1.00     23878
   macro avg       1.00      1.00      1.00     23878
weighted avg       1.00      1.00      1.00     23878



### Decision Tree entropy

In [49]:
from sklearn.tree import DecisionTreeClassifier
dt_e = DecisionTreeClassifier(criterion = "entropy", random_state = 100,max_depth=3, min_samples_leaf=5) 
dt_e.fit(X_train, y_train)
y_pred = dt_e.predict(X_test)

# calculate accuracy
print("\n\nConfusion Matrix: \n\n", confusion_matrix(y_test, y_pred))
print ("\n\nAccuracy : \n\n", accuracy_score(y_test,y_pred)*100)
print("\n\nReport : \n\n", classification_report(y_test, y_pred))



Confusion Matrix: 

 [[14934     0]
 [    0  8944]]


Accuracy : 

 100.0


Report : 

               precision    recall  f1-score   support

           0       1.00      1.00      1.00     14934
           1       1.00      1.00      1.00      8944

   micro avg       1.00      1.00      1.00     23878
   macro avg       1.00      1.00      1.00     23878
weighted avg       1.00      1.00      1.00     23878



### SVM (Support Vector Machine)

### Naive Bayes

In [51]:
from sklearn.naive_bayes import GaussianNB 
gnb = GaussianNB() 
gnb.fit(X_train, y_train) 

y_pred = gnb.predict(X_test)

# calculate accuracy
print("\n\nConfusion Matrix: \n\n", confusion_matrix(y_test, y_pred))
print ("\n\nAccuracy : \n\n", accuracy_score(y_test,y_pred)*100)
print("\n\nReport : \n\n", classification_report(y_test, y_pred))



Confusion Matrix: 

 [[14764   170]
 [    0  8944]]


Accuracy : 

 99.2880475751738


Report : 

               precision    recall  f1-score   support

           0       1.00      0.99      0.99     14934
           1       0.98      1.00      0.99      8944

   micro avg       0.99      0.99      0.99     23878
   macro avg       0.99      0.99      0.99     23878
weighted avg       0.99      0.99      0.99     23878



### KNN (K- Nearest Neighbors)

In [46]:
from sklearn.neighbors import KNeighborsClassifier 
knn = KNeighborsClassifier(n_neighbors=3) 
knn.fit(X_train, y_train) 
y_pred = knn.predict(X_test)

# calculate accuracy
print("\n\nConfusion Matrix: \n\n", confusion_matrix(y_test, y_pred))
print ("\n\nAccuracy : \n\n", accuracy_score(y_test,y_pred)*100)
print("\n\nReport : \n\n", classification_report(y_test, y_pred))



Confusion Matrix: 

 [[14856    78]
 [  747  8197]]


Accuracy : 

 96.54493676187286


Report : 

               precision    recall  f1-score   support

           0       0.95      0.99      0.97     14934
           1       0.99      0.92      0.95      8944

   micro avg       0.97      0.97      0.97     23878
   macro avg       0.97      0.96      0.96     23878
weighted avg       0.97      0.97      0.97     23878



### K-Means

# Gradient Boosting & AdaBoost

In [30]:
from sklearn.ensemble import AdaBoostClassifier

# Instantiate model and fit to data
lr = [0.05, 0.25, 0.50, 0.75, 1.0]

for learning_rate in lr:
    clf = AdaBoostClassifier(n_estimators=50, learning_rate = learning_rate)
    clf.fit(X, y)
    
    print("Learning rate: ", learning_rate)
    print("Accuracy score (training): {0:.3f}".format(clf.score(X, y)))
    print("Accuracy score (validation): {0:.3f}".format(clf.score(X, y)))
    
# Make predictions and store in 'Survived' column of df_test
clf = AdaBoostClassifier(n_estimators=50, learning_rate=1)
clf.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

# calculate accuracy
print("\n\nConfusion Matrix: \n\n", confusion_matrix(y_test, y_pred))
print ("\n\nAccuracy : \n\n", accuracy_score(y_test,y_pred)*100)
print("\n\nReport : \n\n", classification_report(y_test, y_pred))

Learning rate:  0.05
Accuracy score (training): 1.000
Accuracy score (validation): 1.000
Learning rate:  0.25
Accuracy score (training): 1.000
Accuracy score (validation): 1.000
Learning rate:  0.5
Accuracy score (training): 1.000
Accuracy score (validation): 1.000
Learning rate:  0.75
Accuracy score (training): 1.000
Accuracy score (validation): 1.000
Learning rate:  1.0
Accuracy score (training): 1.000
Accuracy score (validation): 1.000


Confusion Matrix: 

 [[14931     3]
 [  259  8685]]


Accuracy : 

 98.90275567467963


Report : 

               precision    recall  f1-score   support

           0       0.98      1.00      0.99     14934
           1       1.00      0.97      0.99      8944

   micro avg       0.99      0.99      0.99     23878
   macro avg       0.99      0.99      0.99     23878
weighted avg       0.99      0.99      0.99     23878



In [50]:
import xgboost as xgb 
my_model = xgb.XGBClassifier() 
my_model.fit(X_train, y_train) 

y_pred = my_model.predict(X_test)

# calculate accuracy
print("\n\nConfusion Matrix: \n\n", confusion_matrix(y_test, y_pred))
print ("\n\nAccuracy : \n\n", accuracy_score(y_test,y_pred)*100)
print("\n\nReport : \n\n", classification_report(y_test, y_pred))



Confusion Matrix: 

 [[14934     0]
 [    0  8944]]


Accuracy : 

 100.0


Report : 

               precision    recall  f1-score   support

           0       1.00      1.00      1.00     14934
           1       1.00      1.00      1.00      8944

   micro avg       1.00      1.00      1.00     23878
   macro avg       1.00      1.00      1.00     23878
weighted avg       1.00      1.00      1.00     23878

